src/HLLD.hpp

```diff
@@ -0,0 +1,346 @@
+	u_star_L.mx = spds[0] * (u_star_L.mx - u_L.mx);
+	u_star_L.my = spds[0] * (u_star_L.my - u_L.my);
+	u_star_L.mz = spds[0] * (u_star_L.mz - u_L.mz);
+	u_star_L.E = spds[0] * (u_star_L.E - u_L.E);
+	u_star_L.by = spds[0] * (u_star_L.by - u_L.by);
+	u_star_L.bz = spds[0] * (u_star_L.bz - u_L.bz);
+
+	u_dstar_R.rho = spds[3] * (u_dstar_R.rho - u_star_R.rho);
+	u_dstar_R.mx = spds[3] * (u_dstar_R.mx - u_star_R.mx);
+	u_dstar_R.my = spds[3] * (u_dstar_R.my - u_star_R.my);
+	u_dstar_R.mz = spds[3] * (u_dstar_R.mz - u_star_R.mz);
+	u_dstar_R.E = spds[3] * (u_dstar_R.E - u_star_R.E);
+	u_dstar_R.by = spds[3] * (u_dstar_R.by - u_star_R.by);
+	u_dstar_R.bz = spds[3] * (u_dstar_R.bz - u_star_R.bz);
+
+	u_star_R.rho = spds[4] * (u_star_R.rho - u_R.rho);
+	u_star_R.mx = spds[4] * (u_star_R.mx - u_R.mx);
+	u_star_R.my = spds[4] * (u_star_R.my - u_R.my);
+	u_star_R.mz = spds[4] * (u_star_R.mz - u_R.mz);
+	u_star_R.E = spds[4] * (u_star_R.E - u_R.E);
+	u_star_R.by = spds[4] * (u_star_R.by - u_R.by);
+	u_star_R.bz = spds[4] * (u_star_R.bz - u_R.bz);
+
+	if (spds[0] >= 0.0) {
+		// return u_L if flow is supersonic
+		f_x.rho = f_L.rho;
+		f_x.mx = f_L.mx;
+		f_x.my = f_L.my;
+		f_x.mz = f_L.mz;
+		f_x.E = f_L.E;
+		f_x.by = f_L.by;
+		f_x.bz = f_L.bz;
+	} else if (spds[4] <= 0.0) {
+		// return u_R if flow is supersonic
+		f_x.rho = f_R.rho;
+		f_x.mx = f_R.mx;
+		f_x.my = f_R.my;
+		f_x.mz = f_R.mz;
+		f_x.E = f_R.E;
+		f_x.by = f_R.by;
+		f_x.bz = f_R.bz;
+	} else if (spds[1] >= 0.0) {
+		// return u_star_L
+		f_x.rho = f_L.rho + u_star_L.rho;
+		f_x.mx = f_L.mx + u_star_L.mx;
+		f_x.my = f_L.my + u_star_L.my;
+		f_x.mz = f_L.mz + u_star_L.mz;
+		f_x.E = f_L.E + u_star_L.E;
+		f_x.by = f_L.by + u_star_L.by;
+		f_x.bz = f_L.bz + u_star_L.bz;
+	} else if (spds[2] >= 0.0) {
+		// return u_dstar_L
+		f_x.rho = f_L.rho + u_star_L.rho + u_dstar_L.rho;
+		f_x.mx = f_L.mx + u_star_L.mx + u_dstar_L.mx;
+		f_x.my = f_L.my + u_star_L.my + u_dstar_L.my;
+		f_x.mz = f_L.mz + u_star_L.mz + u_dstar_L.mz;
+		f_x.E = f_L.E + u_star_L.E + u_dstar_L.E;
+		f_x.by = f_L.by + u_star_L.by + u_dstar_L.by;
+		f_x.bz = f_L.bz + u_star_L.bz + u_dstar_L.bz;
+	} else if (spds[3] > 0.0) {
+		// return u_dstar_R
+		f_x.rho = f_R.rho + u_star_R.rho + u_dstar_R.rho;
+		f_x.mx = f_R.mx + u_star_R.mx + u_dstar_R.mx;
+		f_x.my = f_R.my + u_star_R.my + u_dstar_R.my;
+		f_x.mz = f_R.mz + u_star_R.mz + u_dstar_R.mz;
+		f_x.E = f_R.E + u_star_R.E + u_dstar_R.E;
+		f_x.by = f_R.by + u_star_R.by + u_dstar_R.by;
+		f_x.bz = f_R.bz + u_star_R.bz + u_dstar_R.bz;
+	} else {
+		// return u_star_R
+		f_x.rho = f_R.rho + u_star_R.rho;
+		f_x.mx = f_R.mx + u_star_R.mx;
+		f_x.my = f_R.my + u_star_R.my;
+		f_x.mz = f_R.mz + u_star_R.mz;
+		f_x.E = f_R.E + u_star_R.E;
+		f_x.by = f_R.by + u_star_R.by;
+		f_x.bz = f_R.bz + u_star_R.bz;
+	}
+
+	F_hydro = {f_x.rho, f_x.mx, f_x.my, f_x.mz, f_x.E, 0}; // TODO(neco): Eint=0 for now. Also need to add pscalars.
+	Efield_y = -f_x.by;
+	Efield_z = f_x.bz;
```

Comment:
My preference would be to delete the Efield_y and Efield_z variables, since they are only used by the GS05/CT-Contact scheme.

Reply:
I don't think that is right. They are used in the LD04 method, also. See for example the 1D update step (near the start of the function) in `ComputeCornerE_UCT4` (lines 245-248).

https://github.com/PrincetonUniversity/athena/blob/mhd4_3D/src/field/calculate_corner_e.cpp

Reply:
That is only done for 1D and 2D for the components that are not actually affected by the emf update. In 3D, the electric fields from the Riemann solver are never used.

Reply:
In the 1D and 2D cases, I am almost certain that the electric fields can just be copied without involving the Riemann solver at all.